### ISSUE TITLE
feature: let twisted.conch.ssh.transport supported ssh version to be overriden

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4428]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alanfranzoni reported|
|-|-|
|Trac ID|trac#4428|
|Type|enhancement|
|Created|2010-05-11 10:06:03Z|
|Branch|https://github.com/twisted/twisted/tree/conch-supported-versions-check-4428|

Currently, twisted.conch.ssh.transport ssh protocol supported versions are hardcoded into the dataReceived method, and cannot be overriden. While this isn't usually a problem, some half-baked ssh implementations do sometimes return a strange version number (notably, 2.99) even though the protocol version is actually 2.0.

While this shouldn't be hardcoded in the transport, since it's actually a protocol glitch, this patch allows SSHTransportBase subclassers to override supported version numbers without rewriting the dataReceived method, if this is necessary.

Attachments:

* [4428.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ff/8ff2933e16c95ef5fa35bc0106f5d7ab5fd33957/00049523857a6b77dddfa148771a1b3520c8b543.patch) (2031 bytes) - added by alanfranzoni on 2010-05-11 10:14:37Z - 
* [4428-bis.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ff/8ff2933e16c95ef5fa35bc0106f5d7ab5fd33957/b91dbea7f011d24799b28f653563b209c28199bf.patch) (4492 bytes) - added by alanfranzoni on 2010-05-12 09:51:08Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4428 4428
type__enhancement enhancement
reporter__alanfranzoni alanfranzoni
priority__normal normal
milestone__ 
branch__branches_conch_supported_versions_check_4428 branches/conch-supported-versions-check-4428
branch_author__alanfranzoni alanfranzoni
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1273572363000000 1273572363000000
changetime__1274461591000000 1274461591000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|alanfranzoni removed owner|
|-|-|

Patch attached. Please note that INotify tests on current twisted trunk actually fail on my machine - but they seem to fail without the patch as well, it seems a totally unrelated issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to alanfranzoni|
|-|-|

Thanks.  Can you also add a unit test for this feature?

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|alanfranzoni removed owner|
|-|-|

I had to patch the whole thing a bit more in order to write a proper unit test. I hope this is fine now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [28914]) Branching to 'conch-supported-versions-check-4428'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [28915]) Apply alanfranzoni's patch

refs [#4428](https://github.com/twisted/twisted/issues/4428)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28963]) fix typo in docstring

refs [#4428](https://github.com/twisted/twisted/issues/4428)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  I checked the patch into a branch, fixed some simple doc issues and some coding standard problems, but the basic change looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [28972]) Merge conch-supported-versions-check-4428

Author: alanfranzoni
Reviewer: exarkun
Fixes: [#4428](https://github.com/twisted/twisted/issues/4428)

Add an overrideable attribute which can be used to control the declared
protocol versions `SSHTransportBase` considers supported.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `reopened`|
|-|-|

(In [28975]) Revert r28972 - doc formatting regression

The new `_unsupportedVersionReceived` method's docstring is not valid epytext.

Reopens [#4428](https://github.com/twisted/twisted/issues/4428)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [28979]) Merge conch-supported-versions-check-4428

Author: alanfranzoni
Reviewer: exarkun
Fixes: [#4428](https://github.com/twisted/twisted/issues/4428)

Add an overrideable attribute which can be used to control the declared
protocol versions `SSHTransportBase` considers supported.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
